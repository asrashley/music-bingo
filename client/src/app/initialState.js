import { initialState as cards } from '../cards/cardsSlice';
import { initialState as games } from '../games/gamesSlice';
import { initialState as tickets } from '../tickets/ticketsSlice';
import { initialState as user } from '../user/userSlice';

export const initialState = Object.freeze({
    cards,
    games,
    tickets,
    user,
});
