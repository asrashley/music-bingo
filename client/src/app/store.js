import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import { connectRouter } from 'connected-react-router';
import { createBrowserHistory } from 'history';
import { reducer as formReducer } from 'redux-form';

import cardsReducer from '../cards/cardsSlice';
import gamesReducer from '../games/gamesSlice';
import ticketsReducer from '../tickets/ticketsSlice';
import userReducer from '../user/userSlice';
import adminReducer from '../admin/adminSlice';
import usersMiddleware from '../user/userMiddleware';

export const history = createBrowserHistory();

export const store = configureStore({
  middleware: [...getDefaultMiddleware(), usersMiddleware, logger],
  reducer: {
    admin: adminReducer,
    cards: cardsReducer,
    form: formReducer,
    games: gamesReducer,
    tickets: ticketsReducer,
    router: connectRouter(history),
    user: userReducer,
  },
});
