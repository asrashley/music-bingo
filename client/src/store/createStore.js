import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import { connectRouter } from 'connected-react-router';

import adminReducer from '../admin/adminSlice';
import directoriesReducer from '../directories/directoriesSlice';
import gamesReducer from '../games/gamesSlice';
import messagesReducer from '../messages/messagesSlice';
import settingsReducer from '../settings/settingsSlice';
import systemReducer from '../system/systemSlice';
import ticketsReducer from '../tickets/ticketsSlice';
import userReducer from '../user/userSlice';
import usersMiddleware from '../user/userMiddleware';
import messagesMiddleware from '../messages/messagesMiddleware';

import { history } from './history';

export function createStore(preloadedState, logging) {
  const middleware = [...getDefaultMiddleware(), usersMiddleware, messagesMiddleware];
  if (logging === undefined) {
    logging = (process.env.NODE_ENV === 'development');
  }
  if (logging) {
    middleware.push(logger);
  }
  return configureStore({
    middleware,
    reducer: {
      admin: adminReducer,
      directories: directoriesReducer,
      games: gamesReducer,
      messages: messagesReducer,
      settings: settingsReducer,
      system: systemReducer,
      tickets: ticketsReducer,
      router: connectRouter(history),
      user: userReducer,
    },
    preloadedState
  });
}
