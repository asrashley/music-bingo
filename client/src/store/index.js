export { createStore } from './createStore';
export { initialState } from './initialState';
export { history } from './history';
