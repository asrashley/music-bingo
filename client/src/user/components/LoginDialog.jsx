import React from 'react';
import PropTypes from 'prop-types';
import log from 'loglevel';

import { createGuestAccount, loginUser } from '../userSlice';
import { LoginDialogForm } from './LoginDialogForm';
import { UserPropType } from '../types/User';

import '../styles/user.scss';

class LoginDialog extends React.Component {
  static propTypes = {
    user: UserPropType.isRequired,
    dispatch: PropTypes.func.isRequired,
    onCancel: PropTypes.func.isRequired,
    onSuccess: PropTypes.func.isRequired,
    backdrop: PropTypes.bool,
  };

  state = {
    alert: null,
    lastUpdated: 0,
  };
  mounted = false;

  componentDidMount() {
    this.mounted = true;
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  handleSubmit = ({ password, username, rememberme }) => {
    const { dispatch } = this.props;
    return dispatch(loginUser({ password, username, rememberme }))
      .then(this.submitResponse);
  };

  submitResponse = (result) => {
    if (!this.mounted) {
      return true;
    }
    const { onSuccess } = this.props;
    const { accessToken, error, status } = result.payload ? result.payload : result;
    if (error === undefined && accessToken !== undefined) {
      onSuccess();
      return true;
    }
    this.failedLogin(status, error);
    const errs = [];
    for (let name in error) {
      if (error[name]) {
        errs.push({
          type: "validate",
          message: error[name],
          name,
        });
      }
    }
    if (errs.length === 0) {
      return (error);
    }
    return errs;
  };

  failedLogin(status, err) {
    const lastUpdated = Date.now();
    log.debug(`login failed ${status} ${err}`);
    if (status !== undefined && status >= 500) {
      this.setState({
        alert: "There is a problem with the server. Please try again later",
        lastUpdated,
      });
    } else {
      this.setState({
        alert: "Username or password is incorrect",
        lastUpdated,
      });
    }
  };

  playAsGuest = () => {
    const { dispatch, user } = this.props;
    dispatch(createGuestAccount(user.guest.token));
  };

  render() {
    const { backdrop, user, onCancel } = this.props;
    let { alert, lastUpdated } = this.state;
    let className = "login-form";
    if (user.isisFetching === true) {
      className += " loading";
    }
    if (!alert && user.error && lastUpdated < user.lastUpdated) {
      alert = user.error;
    }
    return (
      <div >
        <LoginDialogForm alert={alert} onSubmit={this.handleSubmit} onCancel={onCancel}
          className={className} user={user} playAsGuest={this.playAsGuest} />
        {backdrop === true && <div className="modal-backdrop fade show"></div>}
      </div>
    );
  }
};

export { LoginDialog };
