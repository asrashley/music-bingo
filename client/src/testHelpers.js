import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { ConnectedRouter } from 'connected-react-router';
import log from 'loglevel';
import { screen, fireEvent } from '@testing-library/react';
import waitForExpect from 'wait-for-expect';

import { DisplayDialog } from './components/DisplayDialog';
import { createStore } from './store/createStore';
import { history } from './store/history';

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    store = createStore(preloadedState),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return (<Provider store={store}>
      <ConnectedRouter history={history}>
        <DisplayDialog>
          {children}
        </DisplayDialog>
      </ConnectedRouter>
    </Provider>);
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

const protectedRoutes = {
  '/api/directory': true,
  '/api/game': true,
  '/api/settings': false,
  '/api/user': true,
};

export function jsonResponse(payload, status = 200) {
  const body = JSON.stringify(payload);
  return {
    body,
    status,
    headers: {
      'Content-Type': 'application/json',
      'Content-Length': body.length
    }
  };
};

export function installFetchMocks(fetchMock, {
  loggedIn = false,
  currentAccessToken = 1,
  refreshToken = "refresh.token"
} = {}) {
  const responseModifiers = {};
  let serverStatus = null;
  const apiRequest = async (url, opts) => {
    log.trace(`apiRequest ${url}`);
    if (serverStatus !== null) {
      return serverStatus;
    }
    if (protectedRoutes[url]) {
      const { headers } = opts;
      log.trace(`Authorization = headers?.Authorization`);
      const bearer = `Bearer ${getAccessToken()}`;
      if (headers?.Authorization !== bearer) {
        log.debug(`Invalid access token "${headers?.Authorization}" !== "${bearer}"`);
        return 401;
      }
    }
    const filename = url.replace(/^\/api/, './fixtures');
    log.trace(`load ${filename}`);
    const data = await import(`${filename}.json`);
    log.trace(`${filename} = ${Object.keys(data['default']).join(',')}`);
    if (url === '/api/settings' && !loggedIn) {
      data['default'] = {
        privacy: data['default'].privacy
      };
    }
    const modFn = responseModifiers[url];
    if (modFn !== undefined) {
      return jsonResponse(modFn(url, data['default']));
    }
    return jsonResponse(data['default']);
  };
  const getAccessToken = () => `access.token.${currentAccessToken}`;
  const refreshAccessToken = (url, opts) => {
    currentAccessToken++;
    log.trace(`refreshAccessToken ${currentAccessToken}`);
    return {
      'accessToken': getAccessToken()
    };
  };
  const checkUser = async (url, opts) => {
    log.debug(`checkUser loggedIn=${loggedIn}`);
    if (serverStatus !== null) {
      log.debug(`checkUser serverStatus=${serverStatus}`);
      return serverStatus;
    }
    if (!loggedIn) {
      return 401;
    }
    const data = await import('./fixtures/user.json');
    data['default'].refreshToken = refreshToken;
    data['default'].accessToken = getAccessToken();
    return jsonResponse(data['default']);
  };
  const userDatabase = [{
    username: 'user',
    email: 'a.user@example.tld',
    password: 'mysecret'
  }];
  const loginUser = async (url, opts) => {
    if (serverStatus !== null) {
      log.debug(`loginUser status=${serverStatus}`);
      return serverStatus;
    }
    const { username, password } = JSON.parse(opts.body);
    const userIsValid = userDatabase.some(item => (
      item.username === username && item.password === password));
    log.debug(`loginUser: username="${username}" password="${password}" valid=${userIsValid}`);
    if (!userIsValid) {
      return 401;
    }
    const data = await import('./fixtures/user.json');
    data['default'].refreshToken = refreshToken;
    data['default'].accessToken = getAccessToken();
    loggedIn = true;
    return jsonResponse(data['default']);
  };
  const logoutUser = (url, opts) => {
    if (serverStatus !== null) {
      log.trace(`logoutUser status=${serverStatus}`);
      return serverStatus;
    }
    log.trace('logoutUser');
    loggedIn = false;
    return jsonResponse('Logged out');
  };
  const checkIfUserExists = (url, opts) => {
    if (serverStatus !== null) {
      log.debug(`checkIfUserExists status=${serverStatus}`);
      return serverStatus;
    }
    const { username, email } = JSON.parse(opts.body);
    const response = {
      "username": false,
      "email": false
    };
    userDatabase.forEach((item) => {
      if (item.username === username) {
        response.username = true;
      }
      if (item.email === email) {
        response.email = true;
      }
    });
    log.debug(`checkIfUserExists username=${username} email=${email} response=${JSON.stringify(response)}`);
    return jsonResponse(response);
  };

  fetchMock.config.fallbackToNetwork = false;
  fetchMock.config.warnOnFallback = true;
  log.debug(`installFetchMocks ${loggedIn}`);
  fetchMock
    .get('/api/directory', apiRequest)
    .get(/api\/directory\/\d+/, apiRequest)
    .get('/api/games', apiRequest)
    .get(/api\/game\/\d+/, apiRequest)
    .post('/api/refresh', refreshAccessToken)
    .get('/api/settings', apiRequest)
    .get('/api/user', checkUser)
    .post('/api/user', loginUser)
    .delete('/api/user', logoutUser)
    .post('/api/user/check', checkIfUserExists)
    .get('/api/users', apiRequest)
    .get('/api/user/guest', apiRequest);

  return {
    getAccessToken,
    getRefreshToken: () => refreshToken,
    isLoggedIn: () => loggedIn,
    addUser: (user) => userDatabase.push(user),
    setResponseModifier: (url, fn) => {
      responseModifiers[url] = fn;
    },
    setServerStatus: (code) => {
      serverStatus = code;
    },
    jsonResponse
  };
}

export async function setFormFields(fields) {
  fields.forEach(({ label, value, exact }) => {
    if (exact === undefined) {
      exact = true;
    }
    fireEvent.input(screen.getByLabelText(label, { exact }), {
      target: {
        value
      }
    });
  });
  await waitForExpect(() => {
    const last = fields[fields.length - 1];
    const { value } = document.querySelector(`input[name="${last.label}"`);
    expect(value).toBe(fields[fields.length - 1].value);
  });
}

