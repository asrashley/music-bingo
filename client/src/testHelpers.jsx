import { render, queries, buildQueries, screen, fireEvent, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import PropTypes from 'prop-types';
import { ReduxRouter } from '@lagunovsky/redux-react-router';
import log from 'loglevel';
import { Readable } from "stream";
import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history';

import { DisplayDialog } from './components/DisplayDialog';
import { createStore } from './store/createStore';
import fetchMockHandlerTest from 'fetch-mock/esm/client';
import jestify from 'fetch-mock-jest/jestify';

import userData from './fixtures/user.json';

export const fetchMock = jestify(fetchMockHandlerTest);

/* custom query that looks for a "data-last-update" attribute in the container and compares that to
  the value provided to the query. It can be used to check if a component has been updated or
  to wait for it to re-render
  */
const queryAllByLastUpdate = (container, lastUpdate, options = {}) => {
  const { comparison } = {
    comparison: 'equals',
    ...options
  };
  log.trace(`queryAllByLastUpdate lastUpdate=${lastUpdate} comparison=${comparison} container="${container.nodeName}.${container.className}" ${container.id}`);
  return Array.from(container.querySelectorAll('[data-last-update]'))
    .filter((elt) => {
      const update = parseInt(elt.dataset.lastUpdate, 10);
      log.trace(`${elt.nodeName}.${elt.className} lastUpdate = ${update}`);
      if (isNaN(update)) {
        return false;
      }
      switch (comparison) {
        case 'equals':
          return update === lastUpdate;
        case 'greaterThan':
          return update > lastUpdate;
        case 'greaterThanOrEquals':
          return update >= lastUpdate;
        case 'lessThan':
          return update < lastUpdate;
        case 'lessThanOrEquals':
          return update <= lastUpdate;
        default:
          return true;
      }
    });
};
export const [queryByLastUpdate, getAllLastUpdate, getByLastUpdate, findAllLastUpdate, findByLastUpdate] = buildQueries(
  queryAllByLastUpdate,
  (container, selector) => `Found multiple elements from ${container} with last update selector: ${selector}`,
  (container, selector, opts = {}) => {
    const { comparison = 'equals' } = opts;
    return (`Unable to find an element from ${container.nodeName}.${container.className} with last update ${comparison} ${selector}`);
  }
);
const lastUpdatedQueries = { queryByLastUpdate, getAllLastUpdate, getByLastUpdate, findAllLastUpdate, findByLastUpdate };

/* custom query that uses a CSS selector to find elements */
const queryAllBySelector = (container, selector) => {
  log.trace(`queryAllBySelector ${selector}`);
  return Array.from(container.querySelectorAll(selector));
};
const [queryBySelector, getAllBySelector, getBySelector, findAllBySelector, findBySelector] = buildQueries(
  queryAllBySelector,
  (container, selector) => `Found multiple elements from ${container} with selector: ${selector}`,
  (container, selector) => `Unable to find an element from ${container} with selector: ${selector}`,
);
const bySelectorQueries = { queryBySelector, getAllBySelector, getBySelector, findAllBySelector, findBySelector };

const routerSelector = (state) => state.router

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    history = createMemoryHistory(),
    store = createStore(preloadedState),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return (<Provider store={store}>
      <ReduxRouter history={history} routerSelector={routerSelector}>
        <DisplayDialog>
          {children}
        </DisplayDialog>
      </ReduxRouter>
    </Provider>);
  }
  Wrapper.propTypes = {
    children: PropTypes.node,
  };
  const events = userEvent.setup({
    delay: 10,
  });
  return {
    events,
    store,
    ...render(ui, {
      wrapper: Wrapper,
      queries: {
        ...queries,
        ...lastUpdatedQueries,
        ...bySelectorQueries
      },
      ...renderOptions
    })
  };
}

const protectedRoutes = {
  '/api/directory': true,
  '/api/game': true,
  '/api/settings': false,
  '/api/user': true,
};

export function jsonResponse(payload, status = 200) {
  const body = JSON.stringify(payload);
  return {
    body,
    status,
    headers: {
      'Cache-Control': 'max-age = 0, no_cache, no_store, must_revalidate',
      'Content-Type': 'application/json',
      'Content-Length': body.length
    }
  };
}

export class MockResponse extends Response {
  constructor(...args) {
    if (args[0] instanceof ReadableStream) {
      args[0] = Readable.from(args[0]);
    }
    super(...args);
  }
}

export function installFetchMocks(fetchMock, {
  loggedIn = false,
  currentAccessToken = 1,
  refreshToken = "refresh.token"
} = {}) {
  const responseModifiers = {};
  let serverStatus = null;
  const apiRequest = async (url, opts) => {
    log.trace(`apiRequest ${url}`);
    if (serverStatus !== null) {
      return serverStatus;
    }
    if (protectedRoutes[url]) {
      const { headers } = opts;
      log.trace(`Authorization = headers?.Authorization`);
      const bearer = `Bearer ${getAccessToken()}`;
      if (headers?.Authorization !== bearer) {
        log.debug(`Invalid access token "${headers?.Authorization}" !== "${bearer}"`);
        return 401;
      }
    }
    const filename = url.replace(/^\/api/, './fixtures');
    log.trace(`load ${filename}`);
    const data = await import(`${filename}.json`);
    log.trace(`${filename} = ${Object.keys(data['default']).join(',')}`);
    if (url === '/api/settings' && !loggedIn) {
      data['default'] = {
        privacy: data['default'].privacy
      };
    }
    const modFn = responseModifiers[url];
    let responsePayload = data['default'];
    if (modFn !== undefined) {
      log.debug(`modify response for ${url}`);
      responsePayload = modFn(url, responsePayload);
      log.debug(`modify response for ${url} done`);
    }
    return jsonResponse(responsePayload);
  };
  const getAccessToken = () => `access.token.${currentAccessToken}`;
  const refreshAccessToken = () => {
    currentAccessToken++;
    log.trace(`refreshAccessToken ${currentAccessToken}`);
    return {
      'accessToken': getAccessToken()
    };
  };
  const checkUser = async () => {
    log.debug(`checkUser loggedIn=${loggedIn}`);
    if (serverStatus !== null) {
      log.debug(`checkUser serverStatus=${serverStatus}`);
      return serverStatus;
    }
    if (!loggedIn) {
      return 401;
    }
    return jsonResponse({
      ...userData,
      refreshToken,
      accessToken: getAccessToken(),
    });
  };
  const userDatabase = [{
    username: 'user',
    email: 'a.user@example.tld',
    password: 'mysecret'
  }];
  const loginUser = async (url, opts) => {
    if (serverStatus !== null) {
      log.debug(`loginUser status=${serverStatus}`);
      return serverStatus;
    }
    const { username, password } = JSON.parse(opts.body);
    let index = -1;
    let dbEntry;
    userDatabase.forEach((item, i) => {
      if (item.username === username && item.password === password) {
        dbEntry = item;
        index = i;
      }
    });
    log.debug(`loginUser: username="${username}" password="${password}" index=${index}`);
    if (!dbEntry) {
      return 401;
    }
    const user = {
      ...userData,
      pk: index + 2,
      ...dbEntry,
      refreshToken,
      accessToken: getAccessToken()
    };
    delete user.password;
    loggedIn = true;
    return jsonResponse(user);
  };
  const logoutUser = () => {
    if (serverStatus !== null) {
      log.trace(`logoutUser status=${serverStatus}`);
      return serverStatus;
    }
    log.trace('logoutUser');
    loggedIn = false;
    return jsonResponse('Logged out');
  };
  const checkIfUserExists = (url, opts) => {
    if (serverStatus !== null) {
      log.trace(`checkIfUserExists status=${serverStatus}`);
      return serverStatus;
    }
    const { username, email } = JSON.parse(opts.body);
    const response = {
      "username": false,
      "email": false
    };
    userDatabase.forEach((item) => {
      if (item.username === username) {
        response.username = true;
      }
      if (item.email === email) {
        response.email = true;
      }
    });
    log.debug(`checkIfUserExists username=${username} email=${email} response=${JSON.stringify(response)}`);
    return jsonResponse(response);
  };

  Object.assign(fetchMock.config, {
    fallbackToNetwork: false,
    warnOnFallback: true,
    Response: MockResponse,
  });
  log.trace(`installFetchMocks ${loggedIn}`);
  fetchMock
    .get('/api/directory', apiRequest)
    .get(/api\/directory\/\d+/, apiRequest)
    .get('/api/games', apiRequest)
    .get(/api\/game\/\d+/, apiRequest)
    .post('/api/refresh', refreshAccessToken)
    .get('/api/settings', apiRequest)
    .get('/api/user', checkUser)
    .post('/api/user', loginUser)
    .delete('/api/user', logoutUser)
    .post('/api/user/check', checkIfUserExists)
    .get('/api/users', apiRequest)
    .get('/api/user/guest', apiRequest);

  return {
    getAccessToken,
    getRefreshToken: () => refreshToken,
    isLoggedIn: () => loggedIn,
    addUser: (user) => userDatabase.push(user),
    logout: () => loggedIn = false,
    setResponseModifier: (url, fn) => {
      responseModifiers[url] = fn;
    },
    setServerStatus: (code) => {
      serverStatus = code;
    },
    jsonResponse
  };
}

export async function setFormFields(fields, events) {
  await Promise.all(fields.map(async ({ label, value, exact }) => {
    if (exact === undefined) {
      exact = true;
    }
    const elt = await screen.findByLabelText(label, { exact });
    if (events) {
      //await events.click(elt);
      await waitFor(() => events.clear(elt));
      /*
            await waitFor(() => events.clear(elt));
        await waitFor(() => events.type(elt, value));
      }*/
    }
    await waitFor(() => fireEvent.input(elt, {
      target: {
        value
      }
    }));
  }));
  const { label, exact, value } = fields[fields.length - 1];
  const elt = await screen.findByLabelText(label, { exact });
  await waitFor(() => {
    expect(elt.value).toBe(value);
  });
}

