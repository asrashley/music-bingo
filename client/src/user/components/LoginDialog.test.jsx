import React from 'react';
import { fireEvent, screen } from '@testing-library/react';
import log from 'loglevel';
import fetchMock from "fetch-mock-jest";
import waitForExpect from 'wait-for-expect';

import { renderWithProviders, installFetchMocks } from '../../testHelpers';
import { createStore } from '../../store/createStore';
import { initialState } from '../../store/initialState';
import * as user from '../../fixtures/userState.json';

import { LoginDialog } from './LoginDialog';

function setUsernameAndPassword({ username, password }) {
  const userField = screen.getByLabelText("User name or email address", { exact: false });
  fireEvent.input(userField, {
    target: {
      value: username
    }
  });
  const pwdField = screen.getByLabelText('Password', { exact: false });
  fireEvent.input(pwdField, {
    target: {
      value: password
    }
  });
}

describe('LoginDialog component', () => {
  let store;
  let apiMock;

  beforeEach(() => {
    store = createStore({
      ...initialState,
      user
    });
    apiMock = installFetchMocks(fetchMock, { loggedIn: false });
  });

  afterEach(() => {
    fetchMock.mockReset();
    log.resetLevel();
    apiMock = null;
  });

  it('renders from initial state', () => {
    const { dispatch } = store;
    const props = {
      user: initialState.user,
      dispatch,
      onCancel: jest.fn(),
      onSuccess: jest.fn(),
      backdrop: true
    };
    const result = renderWithProviders(<LoginDialog {...props} />, { store });
    result.getByLabelText("User name or email address", { exact: false });
  });

  it('login dialog matches snapshot', () => {
    const { dispatch } = store;
    const props = {
      user,
      dispatch,
      onCancel: jest.fn(),
      onSuccess: jest.fn(),
      backdrop: true
    };
    const { asFragment } = renderWithProviders(<LoginDialog {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('calls onCancel when cancel button is pressed', () => {
    const { dispatch } = store;
    const props = {
      user,
      dispatch,
      onCancel: jest.fn(),
      onSuccess: jest.fn(),
    };
    const result = renderWithProviders(<LoginDialog {...props} />);
    fireEvent.click(result.getByRole('button', { name: "Close" }));
    expect(props.onCancel).toHaveBeenCalledTimes(1);
  });

  it('calls onSuccess when submit is successful', async () => {
    const { dispatch } = store;
    const props = {
      user,
      dispatch,
      onCancel: jest.fn(),
      onSuccess: jest.fn(),
    };
    const expected = {
      username: 'user',
      password: 'mysecret',
      rememberme: false
    };
    renderWithProviders(<LoginDialog {...props} />);
    setUsernameAndPassword(expected);
    fireEvent.submit(screen.getByText('Login'));
    await screen.findByText('Logging in..');
    await waitForExpect(() => {
      expect(props.onSuccess).toHaveBeenCalledTimes(1);
    });
  });

  it('shows error message when username or password is wrong', async () => {
    const { dispatch } = store;
    const props = {
      user,
      dispatch,
      onCancel: jest.fn(),
      onSuccess: jest.fn(),
    };
    const expected = {
      username: 'user',
      password: 'wrongpassword',
      rememberme: false
    };
    //log.setLevel('debug');
    renderWithProviders(<LoginDialog {...props} />);
    setUsernameAndPassword(expected);
    fireEvent.submit(screen.getByText('Login'));
    await screen.findByText("Username or password is incorrect");
  });

  it('shows error message when there is a server fault', async () => {
    const { dispatch } = store;
    const props = {
      user,
      dispatch,
      onCancel: jest.fn(),
      onSuccess: jest.fn(),
    };
    const expected = {
      username: 'user',
      password: 'mysecret',
      rememberme: false
    };
    //log.setLevel('debug');
    renderWithProviders(<LoginDialog {...props} />);
    setUsernameAndPassword(expected);
    apiMock.setServerStatus(500);
    fireEvent.submit(screen.getByText('Login'));
    await screen.findByText("There is a problem with the server. Please try again later");
  });

});
