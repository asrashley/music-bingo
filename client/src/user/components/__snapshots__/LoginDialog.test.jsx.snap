// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginDialog component login dialog matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <form
      class="login-form"
    >
      <div
        class="modal show dialog-active "
        data-last-update="0"
        role="dialog"
        style="display: block;"
        tabindex="-1"
      >
        <div
          class="modal-dialog"
          id="login"
          role="document"
        >
          <div
            class="modal-content"
          >
            <div
              class="modal-header"
            >
              <h5
                class="modal-title"
              >
                Log into Musical Bingo
              </h5>
              <button
                aria-label="Close"
                class="close"
                data-dismiss="modal"
                type="button"
              >
                <span
                  aria-hidden="true"
                >
                  Ã—
                </span>
              </button>
            </div>
            <div
              class="modal-body"
            >
              <div
                class="alert alert-warning"
                role="alert"
              >
                <span
                  class="error-message"
                />
              </div>
              <div
                class="form-group"
              >
                <label
                  for="field-username"
                >
                  User name or email address
                  <span
                    class="required"
                  >
                    *
                  </span>
                </label>
                <input
                  class="form-control username  "
                  id="field-username"
                  name="username"
                  placeholder="User name or email address"
                  type="text"
                />
                <small
                  class="form-text text-muted"
                >
                  This is the name you used when you registered your account
                </small>
              </div>
              <div
                class="form-group"
              >
                <label
                  for="field-password"
                >
                  Password
                  <span
                    class="required"
                  >
                    *
                  </span>
                </label>
                <input
                  class="form-control password  "
                  id="field-password"
                  name="password"
                  placeholder="Password"
                  type="password"
                />
                <small
                  class="form-text text-muted"
                />
              </div>
              <div
                class="form-check"
              >
                <input
                  class="form-check-input"
                  id="rememberMe"
                  name="rememberme"
                  type="checkbox"
                />
                <label
                  class="form-check-label"
                  for="rememberMe"
                >
                  Remember me
                </label>
              </div>
            </div>
            <div
              class="modal-footer "
            >
              <div
                class="row mb-2 mt-3"
              >
                <p
                  class="password-reset col-8"
                >
                  <a
                    href="/user/reset"
                  >
                    Help, I have forgotten my password!
                  </a>
                </p>
                <span
                  class="col-4 pr-4"
                >
                  <button
                    aria-label="Submit"
                    class="btn btn-success btn-lg login-button"
                    type="submit"
                  >
                    Login
                  </button>
                </span>
              </div>
              <div
                class="row border-top pt-3 pb-2 create-account"
              >
                <span
                  class="col-5 text-center"
                >
                  <a
                    class="btn btn-primary register-button"
                    href="/register"
                  >
                    Create an account
                  </a>
                </span>
                <span
                  class="col-7"
                >
                  It is free and we won't pass on your details to anyone else.
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
    <div
      class="modal-backdrop fade show"
    />
  </div>
</DocumentFragment>
`;
